Civi/Gpapi/ContractHelper/Sepa.php

```diff
@@ -55,49 +125,223 @@ protected function getObfuscatedIban() {
-      'membership_payment.defer_payment_start'  => 1,
-      'membership_payment.from_ba'              => \CRM_Contract_BankingLogic::getOrCreateBankAccount(
-        $this->contract['contact_id'],
-        $params['payment_details']['iban']
-      ),
-      'membership_payment.to_ba'                => \CRM_Contract_BankingLogic::getCreditorBankAccount(),
-      'membership_payment.membership_annual'    => number_format($params['amount'] * $params['frequency'], 2),
-      'membership_payment.membership_frequency' => $params['frequency'],
-      'membership_payment.cycle_day'            => $this->getCycleDay($cycle_days, $params),
+
+    $bic = CRM_Utils_Array::value('bic', $payment_details);
+    $contact_id = $params['contac_id'];
+    $iban = $payment_details['iban'];
+    $rcur_opt_val = self::getOptionValue('payment_instrument', 'RCUR');
+
+    $cycle_day = $this->getCycleDay($cycle_days, $params);
+    $frequency = $params['frequency'];
+    $from_ba = CRM_Contract_BankingLogic::getOrCreateBankAccount($contact_id, $iban, $bic);
+    $payment_instrument_id = CRM_Utils_Array::value('payment_instrument', $params, $rcur_opt_val);
+    $to_ba = CRM_Contract_BankingLogic::getCreditorBankAccount();
+
+    // API options
+
+    $action = $this->isActiveContract ? 'update' : 'revive';
+
+    $modify_contract_params = [
+      'action'                                  => $action,
+      'campaign_id'                             => $campaign_id,
       'check_permissions'                       => 0,
-    );
-    $response = civicrm_api3('Contract', 'modify', $contract_modification);
+      'date'                                    => $start_date,
+      'id'                                      => $membership_id,
+      'medium_id'                               => $medium_id,
+      'membership_payment.cycle_day'            => $cycle_day,
+      'membership_payment.defer_payment_start'  => 1,
+      'membership_payment.from_ba'              => $from_ba,
+      'membership_payment.membership_annual'    => $annual_amount,
+      'membership_payment.membership_frequency' => $frequency,
+      'membership_payment.to_ba'                => $to_ba,
+      'membership_type_id'                      => $membership_type_id,
+      'payment_method.adapter'                  => 'sepa_mandate',
+      'payment_method.cycle_day'                => $cycle_day,
+      'payment_method.payment_instrument_id'    => $payment_instrument_id,
+    ];
+
+    $response = civicrm_api3('Contract', 'modify', $modify_contract_params);
+
     civicrm_api3('Contract', 'process_scheduled_modifications', [
-      'id'                => $this->membershipId,
+      'id'                => $membership_id,
       'check_permissions' => 0,
     ]);
-    return reset($response['values'])['change_activity_id'];
+
+    $this->loadContract($membership_id);
+  }
+
+  public function createInitialContribution (array $params) {
```

Comment:
This should never be reached for SEPA, where the first payment will always be processed by Civi. We should probably just throw an exception here (it'll just bubble up to whichever silly API client decided to send us SEPA with `payment_received=1`).